File: tests/algorithms/x/Java/other/least_recently_used.java
public class Main {
    static class LRUCache {
        long max_capacity;
        String[] store;
        LRUCache(long max_capacity, String[] store) {
            this.max_capacity = max_capacity;
            this.store = store;
        }
        LRUCache() {}
        @Override public String toString() {
            return String.format("{'max_capacity': %s, 'store': %s}", String.valueOf(max_capacity), String.valueOf(store));
        }
    }

    static LRUCache lru = null;
    static String r = null;

    static LRUCache new_cache(long n) {
        if ((long)(n) < 0L) {
            throw new RuntimeException(String.valueOf("n should be an integer greater than 0."));
        }
        long cap_1 = (long)((long)(n) == 0L ? 2147483647 : n);
        return new LRUCache(cap_1, new String[]{});
    }

    static String[] remove_element(String[] xs, String x) {
        String[] res = ((String[])(new String[]{}));
        boolean removed_1 = false;
        long i_1 = 0L;
        while ((long)(i_1) < (long)(xs.length)) {
            String v_1 = xs[(int)((long)(i_1))];
            if ((removed_1 == false) && (v_1.equals(x))) {
                removed_1 = true;
            } else {
                res = ((String[])(java.util.stream.Stream.concat(java.util.Arrays.stream(res), java.util.Arrays.stream(new String[]{v_1})).toArray(String[]::new)));
            }
            i_1 = (long)((long)(i_1) + 1L);
        }
        return res;
    }

    static LRUCache refer(LRUCache cache, String x) {
        String[] store = ((String[])(cache.store));
        boolean exists_1 = false;
        long i_3 = 0L;
        while ((long)(i_3) < (long)(store.length)) {
            if ((store[(int)((long)(i_3))].equals(x))) {
                exists_1 = true;
            }
            i_3 = (long)((long)(i_3) + 1L);
        }
        if (exists_1) {
            store = ((String[])(remove_element(((String[])(store)), x)));
        } else         if ((long)(store.length) == (long)(cache.max_capacity)) {
            String[] new_store_1 = ((String[])(new String[]{}));
            long j_1 = 0L;
            while ((long)(j_1) < (long)((long)(store.length) - 1L)) {
                new_store_1 = ((String[])(java.util.stream.Stream.concat(java.util.Arrays.stream(new_store_1), java.util.Arrays.stream(new String[]{store[(int)((long)(j_1))]})).toArray(String[]::new)));
                j_1 = (long)((long)(j_1) + 1L);
            }
            store = ((String[])(new_store_1));
        }
        store = ((String[])(java.util.stream.Stream.concat(java.util.Arrays.stream(new String[]{x}), java.util.Arrays.stream(store)).toArray(String[]::new)));
        return new LRUCache(cache.max_capacity, store);
    }

    static void display(LRUCache cache) {
        long i_4 = 0L;
        while ((long)(i_4) < (long)(cache.store.length)) {
            System.out.println(cache.store[(int)((long)(i_4))]);
            i_4 = (long)((long)(i_4) + 1L);
        }
    }

    static String repr_item(String s) {
        boolean all_digits = true;
        long i_6 = 0L;
        while ((long)(i_6) < (long)(_runeLen(s))) {
            String ch_1 = s.substring((int)((long)(i_6)), (int)((long)(i_6))+1);
            if ((ch_1.compareTo("0") < 0) || (ch_1.compareTo("9") > 0)) {
                all_digits = false;
            }
            i_6 = (long)((long)(i_6) + 1L);
        }
        if (all_digits) {
            return s;
        }
        return "'" + s + "'";
    }

    static String cache_repr(LRUCache cache) {
        String res_1 = "LRUCache(" + _p(cache.max_capacity) + ") => [";
        long i_8 = 0L;
        while ((long)(i_8) < (long)(cache.store.length)) {
            res_1 = res_1 + String.valueOf(repr_item(cache.store[(int)((long)(i_8))]));
            if ((long)(i_8) < (long)((long)(cache.store.length) - 1L)) {
                res_1 = res_1 + ", ";
            }
            i_8 = (long)((long)(i_8) + 1L);
        }
        res_1 = res_1 + "]";
        return res_1;
    }
    public static void main(String[] args) {
        lru = new_cache(4L);
        lru = refer(lru, "A");
        lru = refer(lru, "2");
        lru = refer(lru, "3");
        lru = refer(lru, "A");
        lru = refer(lru, "4");
        lru = refer(lru, "5");
        r = String.valueOf(cache_repr(lru));
        System.out.println(r);
        if (!(r.equals("LRUCache(4) => [5, 4, 'A', 3]"))) {
            throw new RuntimeException(String.valueOf("Assertion error"));
        }
    }

    static int _runeLen(String s) {
        return s.codePointCount(0, s.length());
    }

    static String _p(Object v) {
        if (v == null) return "<nil>";
        if (v.getClass().isArray()) {
            if (v instanceof int[]) return java.util.Arrays.toString((int[]) v);
            if (v instanceof long[]) return java.util.Arrays.toString((long[]) v);
            if (v instanceof double[]) return java.util.Arrays.toString((double[]) v);
            if (v instanceof boolean[]) return java.util.Arrays.toString((boolean[]) v);
            if (v instanceof byte[]) return java.util.Arrays.toString((byte[]) v);
            if (v instanceof char[]) return java.util.Arrays.toString((char[]) v);
            if (v instanceof short[]) return java.util.Arrays.toString((short[]) v);
            if (v instanceof float[]) return java.util.Arrays.toString((float[]) v);
            return java.util.Arrays.deepToString((Object[]) v);
        }
        if (v instanceof Double || v instanceof Float) {
            double d = ((Number) v).doubleValue();
            return String.valueOf(d);
        }
        return String.valueOf(v);
    }
}

File: tests/algorithms/x/Pascal/other/least_recently_used.pas
{$mode objfpc}{$modeswitch nestedprocvars}
program Main;
uses SysUtils, Math;
type LRUCache = record
  max_capacity: integer;
  store: array of string;
end;
type StrArray = array of string;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
procedure panic(msg: string);
begin
  writeln(msg);
  halt(1);
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  lru: LRUCache;
  r: string;
  n: integer;
  cache: LRUCache;
  x: string;
  xs: StrArray;
  s: string;
function makeLRUCache(max_capacity: integer; store: StrArray): LRUCache; forward;
function new_cache(n: integer): LRUCache; forward;
function remove_element(xs: StrArray; x: string): StrArray; forward;
function refer(cache: LRUCache; x: string): LRUCache; forward;
procedure display(cache: LRUCache); forward;
function repr_item(s: string): string; forward;
function cache_repr(cache: LRUCache): string; forward;
function makeLRUCache(max_capacity: integer; store: StrArray): LRUCache;
begin
  Result.max_capacity := max_capacity;
  Result.store := store;
end;
function new_cache(n: integer): LRUCache;
var
  new_cache_cap: integer;
begin
  if n < 0 then begin
  panic('n should be an integer greater than 0.');
end;
  if n = 0 then begin
  new_cache_cap := 2147483647;
end else begin
  new_cache_cap := n;
end;
  exit(makeLRUCache(new_cache_cap, StrArray([])));
end;
function remove_element(xs: StrArray; x: string): StrArray;
var
  remove_element_res: array of string;
  remove_element_removed: boolean;
  remove_element_i: integer;
  remove_element_v: string;
begin
  remove_element_res := [];
  remove_element_removed := false;
  remove_element_i := 0;
  while remove_element_i < Length(xs) do begin
  remove_element_v := xs[remove_element_i];
  if (remove_element_removed = false) and (remove_element_v = x) then begin
  remove_element_removed := true;
end else begin
  remove_element_res := concat(remove_element_res, StrArray([remove_element_v]));
end;
  remove_element_i := remove_element_i + 1;
end;
  exit(remove_element_res);
end;
function refer(cache: LRUCache; x: string): LRUCache;
var
  refer_store: array of string;
  refer_exists: boolean;
  refer_i: integer;
  refer_new_store: array of string;
  refer_j: integer;
begin
  refer_store := cache.store;
  refer_exists := false;
  refer_i := 0;
  while refer_i < Length(refer_store) do begin
  if refer_store[refer_i] = x then begin
  refer_exists := true;
end;
  refer_i := refer_i + 1;
end;
  if refer_exists then begin
  refer_store := remove_element(refer_store, x);
end else begin
  if Length(refer_store) = cache.max_capacity then begin
  refer_new_store := [];
  refer_j := 0;
  while refer_j < (Length(refer_store) - 1) do begin
  refer_new_store := concat(refer_new_store, StrArray([refer_store[refer_j]]));
  refer_j := refer_j + 1;
end;
  refer_store := refer_new_store;
end;
end;
  refer_store := concat([x], refer_store);
  exit(makeLRUCache(cache.max_capacity, refer_store));
end;
procedure display(cache: LRUCache);
var
  display_i: integer;
begin
  display_i := 0;
  while display_i < Length(cache.store) do begin
  writeln(cache.store[display_i]);
  display_i := display_i + 1;
end;
end;
function repr_item(s: string): string;
var
  repr_item_all_digits: boolean;
  repr_item_i: integer;
  repr_item_ch: string;
begin
  repr_item_all_digits := true;
  repr_item_i := 0;
  while repr_item_i < Length(s) do begin
  repr_item_ch := s[repr_item_i+1];
  if (repr_item_ch < '0') or (repr_item_ch > '9') then begin
  repr_item_all_digits := false;
end;
  repr_item_i := repr_item_i + 1;
end;
  if repr_item_all_digits then begin
  exit(s);
end;
  exit(('''' + s) + '''');
end;
function cache_repr(cache: LRUCache): string;
var
  cache_repr_res: string;
  cache_repr_i: integer;
begin
  cache_repr_res := ('LRUCache(' + IntToStr(cache.max_capacity)) + ') => [';
  cache_repr_i := 0;
  while cache_repr_i < Length(cache.store) do begin
  cache_repr_res := cache_repr_res + repr_item(cache.store[cache_repr_i]);
  if cache_repr_i < (Length(cache.store) - 1) then begin
  cache_repr_res := cache_repr_res + ', ';
end;
  cache_repr_i := cache_repr_i + 1;
end;
  cache_repr_res := cache_repr_res + ']';
  exit(cache_repr_res);
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  lru := new_cache(4);
  lru := refer(lru, 'A');
  lru := refer(lru, '2');
  lru := refer(lru, '3');
  lru := refer(lru, 'A');
  lru := refer(lru, '4');
  lru := refer(lru, '5');
  r := cache_repr(lru);
  writeln(r);
  if r <> 'LRUCache(4) => [5, 4, ''A'', 3]' then begin
  panic('Assertion error');
end;
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.
